src/metabase/models/collection.clj

```diff
@@ -1492,68 +1572,65 @@
-                                                     :where     [:in :coll.id (into #{} ids-to-fetch)]}))
-                                   {})
-          ;; given a collection, return the collection we will be restoring TO.
-          coll->restore-coll  (fn [coll]
-                                (when (not (or (collection.root/is-root-collection? coll)
-                                               (not (:archived coll))))
-                                  (let [restore-destination (coll-id->restore-coll* (u/the-id coll))]
-                                    (cond
-                                      (:id restore-destination) restore-destination
-                                      (= "/" (:unarchiving_trashed_from_location restore-destination)) collection.root/root-collection
-                                      :else nil))))]
-      (for [coll colls]
+    (let [coll-id->parent-id (into {} (map (fn [{:keys [id parent_id]}]
+                                             [id parent_id])
+                                           (t2/hydrate (filter :archived colls) :parent_id)))
+          parent-ids (keep val coll-id->parent-id)
+          parent-id->archived? (when (seq parent-ids)
+                                 (t2/select-pk->fn :archived :model/Collection :id [:in parent-ids]))]
+      (for [coll colls
+            :let [parent-id (coll-id->parent-id (:id coll))
+                  trashed-directly? (:trashed_directly coll)
+                  parent-archived? (get parent-id->archived? parent-id false)]]
         (cond-> coll
-          (:archived coll) (assoc :can_restore (if-let [restore-destination (coll->restore-coll coll)]
-                                                 (perms/set-has-full-permissions-for-set?
-                                                  @api/*current-user-permissions-set*
-                                                  (perms-for-moving coll restore-destination))
-                                                 false)))))))
+          (:archived coll) (assoc :can_restore (and trashed-directly?
+                                                    (not parent-archived?)
+                                                    (perms/set-has-full-permissions-for-set?
+                                                     @api/*current-user-permissions-set*
+                                                     (perms-for-archiving coll)))))))))
+
+(mi/define-batched-hydration-method can-delete
+  :can_delete
+  "Efficiently hydrate the `:can_delete` of a sequence of items"
+  [items]
+  (when (seq items)
+    (for [item items]
+      (assoc item :can_delete (mi/can-write? item)))))
+
+(mi/define-batched-hydration-method collection-can-delete
+  :collection/can_delete
+  "Right now, you can never permanently delete collections. This is a placeholder in case that changes."
+  [items]
+  (for [item items]
+    (assoc item :can_delete false)))
```

Comment:
Why not just have the `:can_delete` hydration detect whether the item is a collection? Otherwise we're relying on the caller to use the correct hydration method

Reply:
Oh, that's much better - thank you!